File: Lecture/4th/kotlin/deriv.kt
import kotlin.math.sin

fun main(args: Array<String>) {
  println(derivative({x: Double -> sin(x)})(0.0))
}

fun derivative(f: (Double) -> Double): (Double) -> Double {
  val h: Double = 1e-06
  return {x: Double -> (f(x+h) - f(x)) / h}
}

File: Lecture/4th/Code/python/deriv.py
from math import sin, cos

class Derivative:
    def __init__(self, f):
        self.f = f

    def __call__(self, x, h=1e-06):
        return (self.f(x + h) - self.f(x)) / h

def main():
    dsin = Derivative(sin)
    print("dsin=",[dsin(x) for x in range(0, 7)])
    print("cos=", [cos(x) for x in range(0, 7)])
    ddsin = Derivative(dsin)
    print()
    print([ddsin(x) for x in range(0, 7)])
    print([-sin(x) for x in range(0, 7)])

if __name__ == '__main__':
    main()
